job_definitions/rotate_api_tokens.yml

```diff
@@ -0,0 +1,92 @@
+{% set stages = [{'name': 'Add new tokens', 'command': 'add-new'}, {'name': 'Remove old tokens', 'command': 'remove-old'}] %}
+{% set app_groups = [['search-api'], ['api'], ['admin-frontend', 'briefs-frontend', 'brief-responses-frontend', 'buyer-frontend', 'supplier-frontend', 'user-frontend']] %}
+---
+- job:
+    name: rotate-api-tokens
+    display-name: "Rotate API tokens"
+    project-type: pipeline
+    description: Rotate Data and Search API tokens for a given environment.
+    disabled: false
+    concurrent: false
+    parameters:
+      - choice:
+          name: STAGE
+          choices:
+            - preview
+            - staging
+            - production
+    pipeline:
+      script: |
+        def notify_slack(icon, status) {
+          build job: "notify-slack",
+          parameters: [
+            string(name: 'USERNAME', value: "update_api_tokens"),
+            string(name: 'ICON', value: icon),
+            string(name: 'JOB', value: "Rotate API tokens on ${STAGE}"),
+            string(name: 'CHANNEL', value: "#dm-release"),
+            text(name: 'STAGE', value: "${STAGE}"),
+            text(name: 'STATUS', value: status),
+            text(name: 'URL', value: "<${BUILD_URL}consoleFull|${BUILD_DISPLAY_NAME}>")
+          ]
+        }
+
+{% for stage in stages %}
+        stage("{{ stage.name }}") {
+{% if loop.index == 1 %}
+          currentBuild.displayName = "#${BUILD_NUMBER} - ${STAGE}"
+{% endif %}
+          node {
+            try {
+              sh('docker run -e GITHUB_ACCESS_TOKEN digitalmarketplace/scripts scripts/rotate-api-tokens.sh "{{ stage.command }}" "${STAGE}"')
+
+              notify_slack(':spinner:', '{{ stage.name }} - SUCCESS')
+
+            } catch(e) {
+              notify_slack(':kaboom:', '{{ stage.name }} - FAILED')
+              echo "Error caught"
+              currentBuild.result = 'FAILURE'
+              return
+            }
+          }
+        }
+
+        stage('Deploy') {
+          milestone()
+          input(message: "Do not continue until the Pull Request updating API tokens has been merged to master.")
+          milestone()
+
+{% for app_group in app_groups %}
+          parallel(
+{% for app in app_group %}
+            "{{ app }}": {
+              stage("{{ app }}") {
+                build job: "release-app-paas", parameters: [
+                  string(name: "STAGE", value: "${STAGE}"),
+                  string(name: "APPLICATION_NAME", value: "{{ app }}")
+                ]
+              }
+            },
+{% endfor %}
+          )
+{% endfor %}
+        }
+
+{% if loop.index == 1 %}
+        stage("Update Jenkins Config") {
```

Comment:
It feels like as a team we have a poor understanding of why the latest creds need to be in the jenkins repo. I roughly know what it's doing as you showed me the screen earlier but I guess I still don't really know why we have that stuff there and the importance of it being up to date. Could you add something here or in the manual maybe to try and get that knowledge explicit?

Reply:
We don't really have a section on Jenkins in the manual at all. Feels like a page that could do with quite a bit of info.

We do have stuff scattered throughout the manual on Jenkins, but having the core of it in one place would be good.

Reply:
Great point, I'll add a new page in the manual with all our jenkins stuff on my to do list

Reply:
It can be a shared todo list. It's a big page so don't take it all on yourself.